roslibrust/src/ros1/node.rs

```diff
@@ -1,62 +1,361 @@
+        }
+    }
+}
 
 /// Represents a single "real" node, typically only one of these is expected per process
 /// but nothing should specifically prevent that.
 pub struct Node {
     // The xmlrpc client this node uses to make requests to master
-    client: Option<MasterClient>,
-    // Record of what this node publishes
-    publishers: DashMap<String, PublisherHandle>,
+    client: MasterClient,
+    // Server which handles updates from the rosmaster and other ROS nodes
+    _xmlrpc_server: XmlRpcServerHandle,
+    // Receiver for requests to the Node actor
+    node_msg_rx: mpsc::UnboundedReceiver<NodeMsg>,
+    // Map of topic names to the publishing channels associated with the topic
+    publishers: DashMap<String, PublishingChannel>,
     // Record of subscriptions this node has
     subscriptions: DashMap<String, Subscription>,
     // Record of what services this node is serving
     services: DashMap<String, ServiceCallback>,
     // TODO need signal to shutdown xmlrpc server when node is dropped
+    host_addr: Ipv4Addr,
+    hostname: String,
+    node_name: String,
 }
 
 impl Node {
-    // TODO better error
-    async fn new() -> Result<Node, Box<dyn std::error::Error + Send + Sync>> {
-        // We have an initialization loop, because we don't have all the information we
-        // need to create our client until we spawn our server and actually bind our
-        // local TCP port, but our server needs a NodeHandle...
-        // So I'm breaking RAII and assigning client after construction.
-        Ok(Node {
-            client: None,
+    async fn new(
+        master_uri: &str,
+        hostname: &str,
+        node_name: &str,
+        addr: Ipv4Addr,
+    ) -> Result<NodeServerHandle, Box<dyn std::error::Error>> {
+        let (node_sender, node_receiver) = mpsc::unbounded_channel();
+        let node_server_handle = NodeServerHandle {
+            node_server_sender: node_sender,
+        };
+        // Create our xmlrpc server and bind our socket so we know our port and can determine our local URI
+        let xmlrpc_server = XmlRpcServer::new(addr, node_server_handle.clone());
+        let client_uri = format!("http://{hostname}:{}", xmlrpc_server.port());
+
+        let rosmaster_client = MasterClient::new(master_uri, client_uri, node_name).await?;
+        let mut node = Self {
+            client: rosmaster_client,
+            _xmlrpc_server: xmlrpc_server,
+            node_msg_rx: node_receiver,
             publishers: DashMap::new(),
             subscriptions: DashMap::new(),
             services: DashMap::new(),
-        })
-    }
+            host_addr: addr,
+            hostname: hostname.to_owned(),
+            node_name: node_name.to_owned(),
+        };
+
+        tokio::spawn(async move {
+            loop {
+                match node.node_msg_rx.recv().await {
+                    Some(NodeMsg::Shutdown) => {
+                        log::info!("Shutdown requested, shutting down node");
```

Comment:
Not sure if this log message was you or me, but we shouldn't log this if we're not actually going to shutdown node. This looks like a stub? Maybe shove in `unimplemented!`

Reply:
The very next line is a break statement which will exit the loop and cause shutdown.